### PR TITLE
Dive media: Extract thumbnails from videos with ffmpeg

### PR DESCRIPTION
**Pull request by bstoeger**
Signed-off-by: Berthold Stoeger <bstoeger@mail.tuwien.ac.at>

### Describe the pull request:

- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:

This is a test-commit for @atdotde: Rudimentary video-thumbnail extraction with ffmpeg.

### Changes made:

Extract thumbnails from videos via ffmpeg.

### Release note:

Not yet - test-PR
### Documentation change:

Not yet - test-PR

### Mentions:

@atdotde

### CODE REVIEW

**Comment by sfuchs79** [comment]
I tested this and it works nicely here for me under Windows 10.

Simply out of curiosity I even used my "own" cross compiled ffmpeg (version 3.4.2) from MXE. I had to copy ffmpeg.exe and a couple of DLLs to the Subsurface root installation folder.
Just to create no misunderstanding: This is no vote from my side to put this stuff into the Subsurface Windows installer. I can understand that this might not be the preferred solution. In fact it is ~35MB of additional "stuff". Just wanted to tell you that it worked for me in that way :-)

**Comment by dirkhh** [comment]
We can of course *document* how to do this. Here's where you can download things. Here's where you need to copy them.

**Comment by bstoeger** [comment]
> I had to copy ffmpeg.exe and a couple of DLLs to the Subsurface root installation folder.

Cool, then maybe for now we don't even have to use a preferences option. Just tell the user where to put extract the files. For Linux it's simply a matter of having ffmpeg in the PATH, which every sensible distribution should be able to do. For Mac: no idea..!?

Since ffmpeg is GPL + LGPL, I think I could host such an archive.

Note that this is WIP. I want to change the format of the thumbnail files such that for videos we also save the duration (if we know it - see #1494). And it should also be able to contain multiple images (every n-seconds of the video).

**Comment by sfuchs79** [comment]
It may get even better. I'm currently playing with ffmpeg configure options to get a tiny and statically linked ffmpeg.exe which is optimized to only do the specific job we want it to do. Unfortunately with my first try I may have removed too many options - it doesn't work any more (but it is only ~5MB <g>). Could you change your PR in a way that I can see the ffmpeg error message? Currently I only get a "Failed reading ffmpeg output".

**Comment by bstoeger** [comment]
Changed it to print `stderr` in case of failure. The code to print `stdin` is commented out. Perhaps you want to change this to a `printf` so that it is properly formatted. We could also show a `QMessageBox`, perhaps of the collected failures after addition of a bunch of videos.

**Comment by sfuchs79** [comment]
Thanks! With help of the ffmpeg error messages I cross compiled ffmpeg with a few other configurations. I now ended up with a statically linked ffmpeg.exe of about 10MB which does the job of creating thumbnails for me.

Config was (copied from MXE ffmpeg.mk):
         --cross-prefix='$(TARGET)'- \
        --enable-cross-compile \
        --arch=$(firstword $(subst -, ,$(TARGET))) \
        --target-os=mingw32 \
        --prefix='$(PREFIX)/$(TARGET)' \
        --enable-static --disable-shared \
        --enable-small \
        --yasmexe='$(TARGET)-yasm' \
        --disable-debug \
        --disable-pthreads \
        --enable-w32threads \
        --disable-doc \
        --disable-ffplay \
        --disable-ffprobe \
        --disable-ffserver \
        --disable-bsfs \
        --disable-encoders \
        --enable-encoder=mjpeg \
        --disable-filters \
        --enable-filter=scale \
        --disable-muxers \
        --enable-muxer=image2

**Comment by atdotde** [comment]
Of course, you can also use ffmpeg to figure out the duration (or rather its helper ffprobe):

https://superuser.com/questions/650291/how-to-get-video-duration-in-seconds

I don't know if shipping ffmpeg is such a great idea. Why not rather relay on it being there or directing the users to 

https://www.ffmpeg.org/download.html

or using the package manager of their system (on Mac, home-brew happily installs it and places it in the path).

**Comment by bstoeger** [comment]
> Of course, you can also use ffmpeg to figure out the duration (or rather its helper ffprobe)

Yes, but we should strive to give the user the best possible experience without having to install external software as long a the effort stays within reasonable limits. Extracting date and especially duration turns out to be very simple (I'm currently looking at .wmv). I don't want to make ffmpeg a dependency for all video-manipulations. Only for thumbnailing.

**Comment by bstoeger** [comment]
Rebased on master.
Added three preferences fields, which control extraction (quote from commit-message):
- `extract_video_thumbnails` (bool): if true, thumbnails are calculated.
- `extract_video_thumbnail_position` (int 0..100): position in video where thumbnail is fetched.
- `ffmpeg_executable (string)`: path of ffmpeg executable.

If ffmpeg fails to start, thumbnailing is turned off. Preferences-fields not yet wired to frontend!

**Comment by bstoeger** [comment]
Added a "video-file-marker" to mark video thumbnails. The "design" is pure garbage - this will have to be fixed by a more capable person (file: `icons/video_overlay.svg`). Note that the marker is added on-the-fly. The thumbnails on-disk are not messed up. I.e. nothing should be destroyed by the PR as-is.

**Comment by bstoeger** [comment]
Update:
- connected the preferences values to the UI.
- render duration on thumbnails in dive-photo-tab.

@dirkhh, @sfuchs79, @atdotde: From my side this is feature complete, therefore I allowed my self to mark the last commit as fixes #359. This still needs some UI / graphics polishing and there are many options for improvement (@atdotde has the idea of extracting frames every n-seconds; we could also snatch half-a-second of movie, like youtube does). But personally, I'd like to limit my work on this to bug-fixing and polishing, so that I can bother with other things, notably undo.

**Comment by bstoeger** [comment]
I replaced the work-in-progress by the needs-testing label, because from my side, this is ready. 

I would really prefer if this was addressed before the corresponding preferences-code is reworked. Doing it once was painful enough.

**Comment by janiversen** [comment]
If you tell which parts of prefs it is, I will hold thst part back a while.

**Comment by bstoeger** [comment]
```
bool GeneralSettingsObjectWrapper::extractVideoThumbnails() const
{
	return prefs.extract_video_thumbnails;
}
 int GeneralSettingsObjectWrapper::extractVideoThumbnailsPosition() const
{
	return prefs.extract_video_thumbnails_position;
}
 QString GeneralSettingsObjectWrapper::ffmpegExecutable() const
{
	return prefs.ffmpeg_executable;
}
```

That would be very nice - thanks.

File: desktop-widgets/preferences/preferences_defaults.ui [hunk 1]
```diff
@@ -6,8 +6,8 @@
    <rect>
     <x>0</x>
     <y>0</y>
-    <width>561</width>
-    <height>558</height>
+    <width>588</width>
+    <height>600</height>
```
Comment (dirkhh) [inline desktop-widgets/preferences/preferences_defaults.ui]: this always worries me - I assume it's just automatic tool output, but we used to struggle a lot with the preferences window suddenly not fitting in our design goal screen size of 1024x600 and people not being able to change preferences

**Review by dirkhh — COMMENTED** [review]
I haven't tested this, just reviewed the code. That first commit is massive, but it makes sense as is. I worry about the impact on small screens (see my comment), but I don't think that's a show stopper; just something to test in a VM.
Which brings me to what this really needs.
Testing on Windows, Mac, and Linux.
Oh, and an addition to the user manual :-)

File: desktop-widgets/preferences/preferences_defaults.ui (hunk 1 above)
Comment (bstoeger) [reply desktop-widgets/preferences/preferences_defaults.ui]: Yes - that was my fault. Apparently I played with the window-size in Qt Creator. I wonder: Couldn't the window-size be calculated from the preferred sizes of the widgets...?

I removed the gratuitous `leftMargin`, etc. changes. It's really rude of Qt Creator to mangle these things.

**Comment by bstoeger** [comment]
> Which brings me to what this really needs.
> Testing on Windows, Mac, and Linux.

I tested it quite extensively on Kubuntu. Of course, experience shows that testers who are not the author often find problems in use-cases that the author never thought about.

@sfuchs79 tested an older version on Windows.

I don't have access to Mac. So I'm out in that respect. 

> Oh, and an addition to the user manual :-)

Yes, looked into it a few days ago and quickly gave up. There's quite a lot to change!

**Comment by dirkhh** [comment]
Tested on Mac, both with and without `ffmpeg`
VERY nice work @bstoeger

**Pull request merged**